### DISCUSSION TITLE
Two leaders on the same thread partition?

### DISCUSSION THREAD

**Discussion by Glenyz** [openthread/openthread#10696] [Q&A] [answered]
Hi,

I am attempting to get a custom module to act as a joiner device and connect to an existing thread network. I have a commissioner active and it is the leader of the network. However, when I boot up my custom module and I call `otJoinerStart()` and then start the thread, it elects itself as the leader.

How is this possible? Am I misinterpreting something here?

![image](https://github.com/user-attachments/assets/2c33774c-8dbe-4832-9ef0-97d521bbcaf7)

**Comment by jwhui** [comment]
Did you factory reset your joining device to make sure it does not have any existing non-volatile settings stored?

The Commissioning Quick Start guide may be useful:

https://github.com/openthread/openthread/blob/main/src/cli/README_COMMISSIONING.md#start-joiner

**Reply by Glenyz** [reply]
Applying a factory reset still gives the same issue. 

For context I am implemented the joiner with Openthread's API to start the joiner process automatically following the same steps as the guide you sent. The commissioner was set up manually through cli commands and the joiner was turned on after the fact.

**Reply by jwhui** [reply]
A Thread device should not attempt to become a router / leader if it does not have an existing Active Operational Dataset.

Before enabling joiner, can you confirm that there is no existing Active Operational Dataset?

**Reply by Glenyz** [reply]
I tried the following sequence to see if there was a active operational dataset before starting the joiner.
```shell
rtt:~$ rtt:~$ ot thread start
rtt:~$ ot dataset active
Error 23: NotFound
rtt:~$ ot thread start
Done
rtt:~$ ot dataset active
Error 23: NotFound
rtt:~$ ot state
detached
Done
rtt:~$ ot state
leader
Done
rtt:~$ ot dataset active
Active Timestamp: 0
Channel: 11
Channel Mask: 0x07fff800
Ext PAN ID: dead00beef00cafe
Mesh Local Prefix: fdde:ad00:beef:0::/64
Network Key: 3e1c1972ddb48ee39b1ddf20595264f9
Network Name: OpenThread
PAN ID: 0x7d5d
PSKc: 0a8c969fabfddcf67760c7260c8d100f
Security Policy: 672 onrc 0
Done
rtt:~$ 
```

**Reply by jwhui** [reply, marked as answer]
Do you have `OPENTHREAD_CONFIG_OPERATIONAL_DATASET_AUTO_INIT` enabled in your build? This should be disabled by default.

https://github.com/openthread/openthread/blob/2aeb8b833ba760ec29d5f340dd1ce7bcb61c5d56/src/core/config/misc.h#L599

**Reply by Glenyz** [reply]
Edit: Sorry I meant that I didn't enable that config.

**Reply by Glenyz** [reply]
Hi,

I manually disabled it in my config file and the joiner works as expected now thanks!

The config is enabled by default on the NCS in kconfig.features file. Also it doesn't affect it when I manually set up the joiner through cli commands which doesn't really make sense to me.

**Reply by LuDuda** [reply]
@Glenyz indeed in the Zephyr the config is enabled by default (I believe due to legacy reasons).

However I wonder why your device becomes the Leader after you call the `otJoinerStart()`. Basically `ot thread start` should be called only after device successfully joins (receives network credentials) not before. It is enough to enable the IP interface with `ot ifconfig up`. Same applies to the code - where you should enable the Thread after receiving a callback (`otJoinerCallback`) provided to the `otJoinerStart()`.

**Reply by Glenyz** [reply]
Thanks for the insight, In my code I'm not waiting until I receive the callback before I enable the thread which can explain why I'm seeing these behaviours. 
Is there anything I need to configure or change in my program if I add a custom callback function and pass it to `otJoinerStart`? Right now, I'm logging the result when the callback is invoked but it seems that it is never called on. I'm wondering if it's something to do with some configuration that I'm missing again.
